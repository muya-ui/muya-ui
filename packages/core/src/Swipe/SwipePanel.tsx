import React, { useContext, useMemo } from 'react';
import styled, { css } from 'styled-components';

import ImgPoolContext from '../Img/ImgPoolContext';
import forkHandler from '../utils/forkHandler';
import memoForwardRef from '../utils/memoForwardRef';
import useTheme from '../utils/useTheme';
import { ISwipeDirection } from './types';

interface ISwipePanelPureProps {
  $transition: 'on' | 'off';
  $duration: number;
  $direction: ISwipeDirection;
}
type ISwipePanelProps = ISwipePanelPureProps & React.HTMLAttributes<HTMLDivElement>;

const BaseNode = styled.div<ISwipePanelPureProps>`
  position: relative;
  ${props => {
    const { theme, $direction, $transition = 'on', $duration: propDuration } = props;
    const { easing } = theme.transition.spec;
    let panelCss;
    if ($direction === 'horizontal') {
      panelCss = css`
        display: flex;
        width: 100000px;
        height: 100%;
      `;
    }
    let transitionCss;
    if ($transition === 'on') {
      transitionCss = css`
        transition: transform ${propDuration}ms ${easing.easeInOut};
      `;
    }
    return css`
      ${transitionCss}
      ${panelCss}
    `;
  }}
`;

export default memoForwardRef<HTMLDivElement, ISwipePanelProps>((props, ref) => {
  const { onTransitionEnd, ...otherProps } = props;
  const theme = useTheme();

  const pool = useContext(ImgPoolContext);
  const handleTransitionEnd = useMemo(() => {
    return forkHandler(() => {
      pool.throttleCheck({ type: 'scroll' });
    }, onTransitionEnd);
  }, [onTransitionEnd, pool]);
  return <BaseNode theme={theme} {...otherProps} onTransitionEnd={handleTransitionEnd} ref={ref} />;
});
