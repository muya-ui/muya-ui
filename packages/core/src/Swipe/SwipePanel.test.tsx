import React from 'react';
import renderer from 'react-test-renderer';

import SwipePanel from './SwipePanel';

test('测试 SwipePanel horizontal', () => {
  const tree = renderer
    .create(
      <SwipePanel $duration={450} $direction="horizontal" $transition="on">
        <div>AAA</div>
      </SwipePanel>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
  expect(tree).toHaveStyleRule('display', 'flex');
});

test('测试 SwipePanel vertical', () => {
  const tree = renderer
    .create(
      <SwipePanel $duration={450} $direction="vertical" $transition="off">
        <div>AAA</div>
      </SwipePanel>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('测试 SwipePanel duration', () => {
  const tree = renderer
    .create(
      <SwipePanel $duration={4000} $direction="vertical" $transition="on">
        <div>AAA</div>
      </SwipePanel>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
