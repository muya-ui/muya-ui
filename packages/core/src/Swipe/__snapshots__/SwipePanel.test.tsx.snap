// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`测试 SwipePanel duration 1`] = `
.c0 {
  position: relative;
  -webkit-transition: -webkit-transform 4000ms cubic-bezier(0.4,0,0.2,1);
  -webkit-transition: transform 4000ms cubic-bezier(0.4,0,0.2,1);
  transition: transform 4000ms cubic-bezier(0.4,0,0.2,1);
}

<div
  className="c0"
  onTransitionEnd={[Function]}
>
  <div>
    AAA
  </div>
</div>
`;

exports[`测试 SwipePanel horizontal 1`] = `
.c0 {
  position: relative;
  -webkit-transition: -webkit-transform 450ms cubic-bezier(0.4,0,0.2,1);
  -webkit-transition: transform 450ms cubic-bezier(0.4,0,0.2,1);
  transition: transform 450ms cubic-bezier(0.4,0,0.2,1);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100000px;
  height: 100%;
}

<div
  className="c0"
  onTransitionEnd={[Function]}
>
  <div>
    AAA
  </div>
</div>
`;

exports[`测试 SwipePanel vertical 1`] = `
.c0 {
  position: relative;
}

<div
  className="c0"
  onTransitionEnd={[Function]}
>
  <div>
    AAA
  </div>
</div>
`;
